### DISCUSSION TITLE
Can I return arbitrary JSON data from a user script?

### DISCUSSION THREAD

**Discussion by ilmagico** [foxglove/community#860] [Q&A]
I'm learning about user scripts and I saw this [FAQ](https://foxglove.dev/docs/studio/panels/user-scripts#faq) entry called: "Can I return arbitrary JSON data in a message?". It proceeds to answer with an emphatic "Yes! User Scripts supports the json type. You can import it from the ./types module" but the example provided below never imports anything named "json" from the ./types module, nor does it ever return *arbitrary* JSON data. So, is it possible to return arbitrary JSON data, or is the FAQ entry wrong?

Specifically, in my use case, I have a topic that produces a JSON string. I want to parse that and visualize it as a data tree with the "Raw messages" panel. Of course, the schema is not fixed and might be different for every message. I tried to use something like this:
```typescript
type Output = {
  data: any;
};
export default function script(event: Input<"/my_topic">): Output {
  return {data: JSON.parse(event.message.msg)};
};
```
but it complains that it doesn't want `any` there, despite me trying to hide it inside the `Output` type.
So, is it possible to produce arbitrary JSON data? If not, how can I visualize a JSON string in a user friendly way in Foxglove Studio?

**Comment by ilmagico** [comment]
I found a trick to make this happen:
```typescript
type Output = {
  data: string;
};
export default function script(event: Input<"/my_topic">): Output {
  return {data: JSON.parse(event.message.msg) as string};
};
```
This feels hacky and I'm not sure if it's not going to break in the future ...

**Reply by defunctzombie** [reply]
If your `msg` field in your messages is a JSON string then your above code using `JSON.parse` to parse the `msg` into a `data` field that is a javascript object is correct. The one additional item I would call out is the `as string` cast is not right since the parsed data is not a string but an object? Possibly using `unknown` would be more correct if it works in the user script. If not what you have is a reasonable workaround.

**Reply by ilmagico** [reply]
Yes, of course the `as string` cast is incorrect, that's the whole point, it's a hack to trick foxglove into accepting it. Anything else I tried would give errors from foxglove (stuff like it doesn't want `any` to be there). I just tried with `unknown` as the type of the `data` field, and I get: `Unsupported type for member 'data'` from foxglove.

**Reply by defunctzombie** [reply]
I double checked and I think your hack is as good as possible today. User scripts assume you define the output type and don't have any support for "any" or "unknown"

**Reply by ilmagico** [reply]
Thanks @defunctzombie for checking, I'll continue to use the current hack for the time being, until I find a better way.
Anyways, I think the FAQ entry needs to be updated, right? or did I just misread it?

**Reply by defunctzombie** [reply]
Yea - the FAQ is wrong - we will get that fixed.

**Reply by ilmagico** [reply]
Awesome, thanks again!
